packages/common/src/codegen/utils/extractUserTypes.ts

```diff
@@ -0,0 +1,57 @@
+import { parse, visit } from "@solidity-parser/parser";
+import { MUDError } from "../../errors";
+
+export interface SolidityUserDefinedType {
+  typeId: string;
+  internalTypeId: string;
+  importSymbol: string;
+  fromPath: string;
+  isRelativePath: boolean;
+}
+
+/**
+ * Parse the solidity data to extract user-defined type information.
+ * @param data contents of a solidity file with the user types declarations
+ * @param userTypeNames names of the user types to extract
+ */
+export function extractUserTypes(
+  data: string,
+  userTypeNames: string[],
+  fromPath: string
+): Record<string, SolidityUserDefinedType> {
+  const ast = parse(data);
+
+  const isRelativePath = fromPath.indexOf(".") === 0;
```

Comment:
nit, don't need to check the whole string for index of `"."` if it's not at the first position
```suggestion
  const isRelativePath = fromPath.at(0) === ".";
```